File: 0020_Valid_Parentheses/solution.py
class Solution:
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        pile = []
        couple = {')': '(', '}': '{', ']': '['}
        for char in s:
            if char in couple:
                if not pile or pile[-1] != couple[char]:
                    return False
                pile.pop()
            else:
                pile.append(char)
        return not pile

if __name__ == "__main__":
    sol = Solution()
    data = input("Enter a string of parentheses: ")
    result = sol.isValid(data)
    print(f"The string is {'valid' if result else 'invalid'}.")

File: 0020_Valid_Parentheses/solution.cpp
#include <iostream>
#include <stack>
#include <string>
#include <unordered_map>
using namespace std;

class Solution {
    public:
        bool isValid(string s){
            unordered_map<char, char> couple = {
                {')', '('},
                {']', '['},
                {'}', '{'}
            };
            stack<char> pile;
            for (char c: s){
                if (couple.count(c)){
                    if (pile.empty() || pile.top() != couple[c]) {
                        return false;
                    }
                    pile.pop();
                } else {
                    pile.push(c);
                }
            }
            return pile.empty();
        }
};
int main(){
    Solution sol;
    string data;
    cout << "Enter a string of parentheses: ";
    cin >> data;
    if (sol.isValid(data)){
        cout << "The string is valid." << endl;
    } else {
        cout << "The string is invalid." << endl;
    }
    return 0;
}